File: MINDSET_module-main/Claude Code/temp/test_product_distribution.py
#!/usr/bin/env python
"""
Test Script - Product Distribution Dataset Builder
This runs key validation checks on the actual data before full processing
"""

import pandas as pd
import numpy as np
import os

# Configuration
DATA_PATH = r"C:\Users\festeves\OneDrive - RAND Corporation\Courses\Dissertation\3rd Paper\Data"
COST_STRUCTURE_PATH = os.path.join(DATA_PATH, "Cost Structure")
CROSSWALK_FILE = os.path.join(DATA_PATH, "GLORIA-Eora26 - Crosswalk.xlsx")
COUNTRIES = ["BGR", "BLZ", "EGY", "LBY", "MAR", "MEX", "UGA"]

print("="*80)
print("TESTING PRODUCT DISTRIBUTION DATASET BUILDER")
print("="*80)

# ==============================================================================
# TEST 1: Load Crosswalk
# ==============================================================================
print("\nTEST 1: Loading Crosswalk")
print("-"*80)

try:
    crosswalk_wide = pd.read_excel(CROSSWALK_FILE, sheet_name="Eora26 - GLORIA")
    print("SUCCESS: Loaded crosswalk")
    print(f"  Dimensions: {crosswalk_wide.shape[0]} rows × {crosswalk_wide.shape[1]} columns")
    print(f"  Column names: {', '.join(crosswalk_wide.columns[:5])}...")

    # Identify sector columns
    sector_cols = [col for col in crosswalk_wide.columns
                   if col not in ['Products', 'Codes', 'Lfd_Nr']]
    print(f"  Sector columns identified: {len(sector_cols)}")
    print(f"  Sectors: {', '.join(sector_cols[:5])}...")

except Exception as e:
    print(f"ERROR loading crosswalk: {e}")
    exit(1)

# ==============================================================================
# TEST 2: Transform Crosswalk to Long
# ==============================================================================
print("\nTEST 2: Transform Crosswalk to Long Format")
print("-"*80)

try:
    # Melt to long format
    crosswalk_long = crosswalk_wide.melt(
        id_vars=['Products'],
        value_vars=sector_cols,
        var_name='Sector',
        value_name='linked'
    )

    # Keep only links (where value = 1)
    crosswalk_long = crosswalk_long[crosswalk_long['linked'] == 1][['Products', 'Sector']]

    print("SUCCESS: Transformed to long format")
    print(f"  Rows: {len(crosswalk_long)} (Product-Sector pairs)")
    print(f"  Unique products: {crosswalk_long['Products'].nunique()}")
    print(f"  Unique sectors: {crosswalk_long['Sector'].nunique()}")

    print("\nSample Product-Sector links:")
    print(crosswalk_long.head(10).to_string())

    # Count products per sector
    products_per_sector = crosswalk_long.groupby('Sector').size().reset_index(name='n_products')

    print("\nProducts per sector:")
    print(products_per_sector.to_string())

    # Add to crosswalk
    crosswalk_long = crosswalk_long.merge(products_per_sector, on='Sector')

except Exception as e:
    print(f"ERROR transforming crosswalk: {e}")
    import traceback
    traceback.print_exc()
    exit(1)

# ==============================================================================
# TEST 3: Load Distribution Files (Test with 1 country first)
# ==============================================================================
print("\n" + "="*80)
print("TEST 3: Loading Distribution Files")
print("-"*80)

test_country = COUNTRIES[0]
print(f"Testing with: {test_country}")

try:
    file_path = os.path.join(COST_STRUCTURE_PATH, f"cost_str_{test_country}.xlsx")

    if not os.path.exists(file_path):
        print(f"  ERROR: File not found: {file_path}")
        exit(1)

    print(f"  Loading: {os.path.basename(file_path)}")

    test_dist = pd.read_excel(file_path)
    test_dist['Country'] = test_country

    print("SUCCESS: Loaded distribution file")
    print(f"  Dimensions: {test_dist.shape[0]} rows × {test_dist.shape[1]} columns")
    print(f"  Columns: {', '.join(test_dist.columns)}")

    print("\nFirst 10 rows:")
    print(test_dist.head(10).to_string())

    # Check required columns
    required_cols = ['Sector', 'strategy_id', 'value']
    missing = [col for col in required_cols if col not in test_dist.columns]

    if missing:
        print(f"\nERROR: Missing required columns: {', '.join(missing)}")
        exit(1)
    else:
        print("\nSUCCESS: All required columns present")

    # Check unique values
    print(f"\n  Unique strategies: {test_dist['strategy_id'].nunique()}")
    print(f"  Unique sectors: {test_dist['Sector'].nunique()}")
    print(f"  Value range: {test_dist['value'].min():.6f} to {test_dist['value'].max():.6f}")

    # Check sums
    sum_check = test_dist.groupby('strategy_id')['value'].sum().reset_index(name='total')

    print("\n  Sum of values per strategy:")
    print(f"    Mean: {sum_check['total'].mean():.6f}")
    print(f"    Min: {sum_check['total'].min():.6f}")
    print(f"    Max: {sum_check['total'].max():.6f}")

    bad_sums = sum_check[np.abs(sum_check['total'] - 1.0) > 1e-6]
    if len(bad_sums) > 0:
        print(f"\n  WARNING: {len(bad_sums)} strategies don't sum to 1.0:")
        print(bad_sums.head().to_string())
    else:
        print("\n  SUCCESS: All strategies sum to 1.0")

except Exception as e:
    print(f"ERROR loading distribution: {e}")
    import traceback
    traceback.print_exc()
    exit(1)

# ==============================================================================
# TEST 4: Check Sector Name Matching
# ==============================================================================
print("\n" + "="*80)
print("TEST 4: Check Sector Name Matching")
print("-"*80)

try:
    sectors_crosswalk = set(crosswalk_long['Sector'].unique())
    sectors_distribution = set(test_dist['Sector'].unique())

    print(f"Sectors in crosswalk: {len(sectors_crosswalk)}")
    print(f"Sectors in distribution: {len(sectors_distribution)}")

    in_crosswalk_not_dist = sectors_crosswalk - sectors_distribution
    in_dist_not_crosswalk = sectors_distribution - sectors_crosswalk

    if in_crosswalk_not_dist:
        print("\nWARNING: Sectors in crosswalk but NOT in distribution:")
        for sector in sorted(in_crosswalk_not_dist):
            print(f"  - {sector}")

    if in_dist_not_crosswalk:
        print("\nWARNING: Sectors in distribution but NOT in crosswalk:")
        for sector in sorted(in_dist_not_crosswalk):
            print(f"  - {sector}")

    if not in_crosswalk_not_dist and not in_dist_not_crosswalk:
        print("\nSUCCESS: All sector names match perfectly!")

except Exception as e:
    print(f"ERROR checking sector names: {e}")
    import traceback
    traceback.print_exc()

# ==============================================================================
# TEST 5: Test Merge and Allocation (Small Sample)
# ==============================================================================
print("\n" + "="*80)
print("TEST 5: Test Merge and Allocation")
print("-"*80)

try:
    # Test with first strategy only
    test_strategy = test_dist['strategy_id'].iloc[0]
    print(f"Testing with strategy: {test_strategy}")

    test_subset = test_dist[test_dist['strategy_id'] == test_strategy].copy()

    # Merge with crosswalk
    test_merge = test_subset.merge(
        crosswalk_long,
        left_on='Sector',
        right_on='Sector',
        how='left'
    )

    # Remove rows where product is NaN (sectors not in crosswalk)
    test_merge = test_merge[test_merge['Products'].notna()]

    # Calculate product-level values
    test_merge['value_sector'] = test_merge['value']
    test_merge['value_product'] = test_merge['value_sector'] / test_merge['n_products']

    print("SUCCESS: Merged and allocated")
    print(f"  Merged rows: {len(test_merge)}")
    print(f"  Unique products: {test_merge['Products'].nunique()}")

    print("\nSample allocated values:")
    sample = test_merge[['Country', 'strategy_id', 'Products', 'Sector', 'n_products', 'value_product']].head(15)
    print(sample.to_string())

    # Check sum
    total_value_product = test_merge['value_product'].sum()
    print(f"\nTotal Value_Product for this strategy: {total_value_product:.10f}")
    print(f"Should be 1.0. Difference: {abs(total_value_product - 1.0):.10e}")

    if abs(total_value_product - 1.0) < 1e-6:
        print("SUCCESS: Values sum to 1.0!")
    else:
        print("WARNING: Values do NOT sum to 1.0")
        print(f"  Missing: {1.0 - total_value_product:.6f}")

except Exception as e:
    print(f"ERROR during merge/allocation: {e}")
    import traceback
    traceback.print_exc()

# ==============================================================================
# TEST 6: Check Multi-Sector Products
# ==============================================================================
print("\n" + "="*80)
print("TEST 6: Check Multi-Sector Products")
print("-"*80)

try:
    # Check for products appearing multiple times
    multi_check = test_merge.groupby('Products').agg({
        'Sector': lambda x: ' + '.join(x),
        'value_product': 'sum',
        'Country': 'size'
    }).rename(columns={'Country': 'n_sectors'})

    multi_check = multi_check[multi_check['n_sectors'] > 1]

    if len(multi_check) > 0:
        print(f"Found {len(multi_check)} products linked to multiple sectors")
        print(multi_check.head(10).to_string())
        print("\nAfter summing, these products will have higher values")
    else:
        print("No multi-sector products found in this test")

except Exception as e:
    print(f"ERROR checking multi-sector: {e}")
    import traceback
    traceback.print_exc()

# ==============================================================================
# TEST 7: Check All Country Files
# ==============================================================================
print("\n" + "="*80)
print("TEST 7: Check All Country Files")
print("-"*80)

for country in COUNTRIES:
    file_path = os.path.join(COST_STRUCTURE_PATH, f"cost_str_{country}.xlsx")

    if os.path.exists(file_path):
        print(f"  {country}: FOUND")
    else:
        print(f"  {country}: NOT FOUND ({file_path})")

# ==============================================================================
# SUMMARY
# ==============================================================================
print("\n" + "="*80)
print("TEST SUMMARY")
print("="*80)

print("\nAll critical tests completed.")
print("\nNext steps:")
print("  1. Review any warnings above")
print("  2. If all looks good, run the full Rmd file in RStudio/Positron")
print("  3. Check final output for validation")
print("\n" + "="*80)

File: MINDSET_module-main/Claude Code/temp/test_product_distribution.R
# Test Script - Product Distribution Dataset Builder
# This runs the key logic to spot errors before full processing

library(tidyverse)
library(readxl)

# Paths
DATA_PATH <- "C:/Users/festeves/OneDrive - RAND Corporation/Courses/Dissertation/3rd Paper/Data"
COST_STRUCTURE_PATH <- file.path(DATA_PATH, "Cost Structure")
CROSSWALK_FILE <- file.path(DATA_PATH, "GLORIA-Eora26 - Crosswalk.xlsx")
COUNTRIES <- c("BGR", "BLZ", "EGY", "LBY", "MAR", "MEX", "UGA")

cat("="*80, "\n", sep = "")
cat("TESTING PRODUCT DISTRIBUTION DATASET BUILDER\n")
cat("="*80, "\n", sep = "")

# ==============================================================================
# TEST 1: Load Crosswalk
# ==============================================================================
cat("\nTEST 1: Loading Crosswalk\n")
cat("-"*80, "\n", sep = "")

tryCatch({
  crosswalk_wide <- read_excel(CROSSWALK_FILE, sheet = "Eora26 - GLORIA")
  cat("SUCCESS: Loaded crosswalk\n")
  cat("  Dimensions:", nrow(crosswalk_wide), "rows ×", ncol(crosswalk_wide), "columns\n")
  cat("  Column names:", paste(names(crosswalk_wide)[1:5], collapse = ", "), "...\n")

  # Identify sector columns
  product_col <- "Products"
  sector_cols <- setdiff(names(crosswalk_wide), c(product_col, "Codes", "Lfd_Nr"))
  cat("  Sector columns identified:", length(sector_cols), "\n")
  cat("  Sectors:", paste(head(sector_cols, 5), collapse = ", "), "...\n")

}, error = function(e) {
  cat("ERROR loading crosswalk:", e$message, "\n")
  stop()
})

# ==============================================================================
# TEST 2: Transform Crosswalk to Long
# ==============================================================================
cat("\nTEST 2: Transform Crosswalk to Long Format\n")
cat("-"*80, "\n", sep = "")

tryCatch({
  crosswalk_long <- crosswalk_wide %>%
    select(Products, all_of(sector_cols)) %>%
    pivot_longer(
      cols = all_of(sector_cols),
      names_to = "Sector",
      values_to = "linked"
    ) %>%
    filter(linked == 1) %>%
    select(Products, Sector)

  cat("SUCCESS: Transformed to long format\n")
  cat("  Rows:", nrow(crosswalk_long), "(Product-Sector pairs)\n")
  cat("  Unique products:", n_distinct(crosswalk_long$Products), "\n")
  cat("  Unique sectors:", n_distinct(crosswalk_long$Sector), "\n")

  cat("\nSample Product-Sector links:\n")
  print(head(crosswalk_long, 10))

  # Count products per sector
  products_per_sector <- crosswalk_long %>%
    group_by(Sector) %>%
    summarize(n_products = n(), .groups = "drop")

  cat("\nProducts per sector:\n")
  print(products_per_sector)

  # Add to crosswalk
  crosswalk_long <- crosswalk_long %>%
    left_join(products_per_sector, by = "Sector")

}, error = function(e) {
  cat("ERROR transforming crosswalk:", e$message, "\n")
  stop()
})

# ==============================================================================
# TEST 3: Load Distribution Files (Test with 1 country first)
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST 3: Loading Distribution Files\n")
cat("-"*80, "\n", sep = "")

# Function to load one country
load_country_distribution <- function(country) {
  file_path <- file.path(COST_STRUCTURE_PATH, paste0("cost_str_", country, ".xlsx"))

  if (!file.exists(file_path)) {
    cat("  WARNING: File not found:", file_path, "\n")
    return(NULL)
  }

  cat("  Loading:", basename(file_path), "\n")

  df <- read_excel(file_path) %>%
    mutate(Country = country) %>%
    select(Country, everything())

  return(df)
}

# Test with first country
test_country <- COUNTRIES[1]
cat("Testing with:", test_country, "\n")

tryCatch({
  test_dist <- load_country_distribution(test_country)

  if (!is.null(test_dist)) {
    cat("SUCCESS: Loaded distribution file\n")
    cat("  Dimensions:", nrow(test_dist), "rows ×", ncol(test_dist), "columns\n")
    cat("  Columns:", paste(names(test_dist), collapse = ", "), "\n")

    cat("\nFirst 10 rows:\n")
    print(head(test_dist, 10))

    # Check required columns
    required_cols <- c("Sector", "strategy_id", "value")
    missing <- setdiff(required_cols, names(test_dist))

    if (length(missing) > 0) {
      cat("\nERROR: Missing required columns:", paste(missing, collapse = ", "), "\n")
      stop()
    } else {
      cat("\nSUCCESS: All required columns present\n")
    }

    # Check unique values
    cat("\n  Unique strategies:", n_distinct(test_dist$strategy_id), "\n")
    cat("  Unique sectors:", n_distinct(test_dist$Sector), "\n")
    cat("  Value range:", min(test_dist$value), "to", max(test_dist$value), "\n")

    # Check sums
    sum_check <- test_dist %>%
      group_by(strategy_id) %>%
      summarize(total = sum(value), .groups = "drop")

    cat("\n  Sum of values per strategy:\n")
    print(summary(sum_check$total))

    if (any(abs(sum_check$total - 1.0) > 1e-6)) {
      cat("\n  WARNING: Some strategies don't sum to 1.0\n")
      bad <- sum_check %>% filter(abs(total - 1.0) > 1e-6)
      print(head(bad))
    }
  }

}, error = function(e) {
  cat("ERROR loading distribution:", e$message, "\n")
  stop()
})

# ==============================================================================
# TEST 4: Check Sector Name Matching
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST 4: Check Sector Name Matching\n")
cat("-"*80, "\n", sep = "")

tryCatch({
  sectors_crosswalk <- unique(crosswalk_long$Sector)
  sectors_distribution <- unique(test_dist$Sector)

  cat("Sectors in crosswalk:", length(sectors_crosswalk), "\n")
  cat("Sectors in distribution:", length(sectors_distribution), "\n")

  in_crosswalk_not_dist <- setdiff(sectors_crosswalk, sectors_distribution)
  in_dist_not_crosswalk <- setdiff(sectors_distribution, sectors_crosswalk)

  if (length(in_crosswalk_not_dist) > 0) {
    cat("\nWARNING: Sectors in crosswalk but NOT in distribution:\n")
    print(in_crosswalk_not_dist)
  }

  if (length(in_dist_not_crosswalk) > 0) {
    cat("\nWARNING: Sectors in distribution but NOT in crosswalk:\n")
    print(in_dist_not_crosswalk)
  }

  if (length(in_crosswalk_not_dist) == 0 && length(in_dist_not_crosswalk) == 0) {
    cat("\nSUCCESS: All sector names match perfectly!\n")
  }

}, error = function(e) {
  cat("ERROR checking sector names:", e$message, "\n")
})

# ==============================================================================
# TEST 5: Test Merge and Allocation (Small Sample)
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST 5: Test Merge and Allocation\n")
cat("-"*80, "\n", sep = "")

tryCatch({
  # Test with first strategy only
  test_strategy <- test_dist$strategy_id[1]
  cat("Testing with strategy:", test_strategy, "\n")

  test_subset <- test_dist %>%
    filter(strategy_id == test_strategy)

  # Merge with crosswalk
  test_merge <- test_subset %>%
    left_join(
      crosswalk_long,
      by = c("Sector" = "Sector"),
      relationship = "many-to-many"
    ) %>%
    filter(!is.na(Products)) %>%
    mutate(
      value_sector = value,
      value_product = value_sector / n_products
    )

  cat("SUCCESS: Merged and allocated\n")
  cat("  Merged rows:", nrow(test_merge), "\n")
  cat("  Unique products:", n_distinct(test_merge$Products), "\n")

  cat("\nSample allocated values:\n")
  print(head(test_merge %>% select(Country, strategy_id, Products, Sector, n_products, value_product), 15))

  # Check sum
  total_value_product <- sum(test_merge$value_product)
  cat("\nTotal Value_Product for this strategy:", total_value_product, "\n")
  cat("Should be 1.0. Difference:", abs(total_value_product - 1.0), "\n")

  if (abs(total_value_product - 1.0) < 1e-6) {
    cat("SUCCESS: Values sum to 1.0!\n")
  } else {
    cat("WARNING: Values do NOT sum to 1.0\n")
  }

}, error = function(e) {
  cat("ERROR during merge/allocation:", e$message, "\n")
  traceback()
})

# ==============================================================================
# TEST 6: Check Multi-Sector Products
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST 6: Check Multi-Sector Products\n")
cat("-"*80, "\n", sep = "")

tryCatch({
  # Check for products appearing multiple times
  multi_check <- test_merge %>%
    group_by(Products) %>%
    summarize(
      n_sectors = n(),
      sectors = paste(Sector, collapse = " + "),
      total_value = sum(value_product),
      .groups = "drop"
    ) %>%
    filter(n_sectors > 1)

  if (nrow(multi_check) > 0) {
    cat("Found", nrow(multi_check), "products linked to multiple sectors\n")
    print(multi_check)

    cat("\nAfter summing, these products will have higher values\n")
  } else {
    cat("No multi-sector products found in this test\n")
  }

}, error = function(e) {
  cat("ERROR checking multi-sector:", e$message, "\n")
})

# ==============================================================================
# TEST 7: Load All Countries (Quick Check)
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST 7: Check All Country Files\n")
cat("-"*80, "\n", sep = "")

for (country in COUNTRIES) {
  file_path <- file.path(COST_STRUCTURE_PATH, paste0("cost_str_", country, ".xlsx"))

  if (file.exists(file_path)) {
    cat("  ", country, ": FOUND\n")
  } else {
    cat("  ", country, ": NOT FOUND (", file_path, ")\n")
  }
}

# ==============================================================================
# SUMMARY
# ==============================================================================
cat("\n", "="*80, "\n", sep = "")
cat("TEST SUMMARY\n")
cat("="*80, "\n", sep = "")

cat("\nAll critical tests completed.\n")
cat("\nNext steps:\n")
cat("  1. Review any warnings above\n")
cat("  2. If all looks good, run the full Rmd file\n")
cat("  3. Check final output for validation\n")
cat("\n", "="*80, "\n", sep = "")
